File: P_4/Examen_P4/Napakalaki_Ruby_P4/lib/player.rb
# encoding: utf-8

module NapakalakiRuby

require_relative 'treasure'
require_relative 'monster'
require_relative 'treasure_kind'
require_relative 'combat_result'
require_relative 'dice'
require_relative 'card_dealer'
require_relative 'bad_consequence'
require_relative 'numeric_bad_consequence'
require_relative 'specific_bad_consequence'
require_relative 'death_bad_consequence'
    
class Player
    
    @@MAXLEVEL = 10
    
    # Consultores basicos
    attr_reader :name
    attr_reader :level
    attr_reader :dead
    attr_reader :canISteal
    attr_reader :visible_treasures
    attr_reader :hidden_treasures
    attr_writer :enemy, :pending_bc
    
    # Constructor
    def initialize(name)
                
        @name = name
        @level = 1
        @dead = true
        @canISteal = true
        
        @enemy = nil
        @pending_bc = nil
        
        @visible_treasures = Array.new
        @hidden_treasures = Array.new
    end
    
    def self.crear_copia(p)
        @name = p.name
        @level = p.level
        @dead = p.dead
        @canISteal = p.canISteal
        @enemy = p.enemy
        @pending_bc = p.pending_bc
        @visible_treasures = p.visible_treasures
        @hidden_treasures = p.hidden_treasures
    end
    
    def to_s
        "#{self.name}
        \n Nivel: #{@level.to_s}"
#        \n Enemigo: #{@enemy.name}"
    end
    
    public
    
    def self.get_max_level
        return @@MAXLEVEL
    end
    
    def is_dead
        return @dead
    end
    
    def combat_level
        @lvl = level
    end
    
    def bring_to_life
        @dead = false
    end
    
    def increment_levels(l)
        @level += l
    end
    
    def decrement_levels(l)
        @level -= l
        
        if (@level < 1)
            @level = 1
        end
    end
    
    def get_combat_level
        
        combat_level = @level
        
        @visible_treasures.each do |t| 
            combat_level = combat_level + t.bonus
        end
        
        return (combat_level)
    end
    
    def set_pending_badconsequence(bc)
        @pending_badconseuqnece = bc
    end
    
    def apply_prize(current_monster)
        
        n_levels = current_monster.get_levels_gained
        increment_levels(n_levels)
        
        n_treasures = current_monster.get_treasures_gained
        
        if (n_treasures > 0)
            dealer = CardDealer.instance
            
            n_treasures.times do
                t = dealer.next_treasure
                @hidden_treasures << t
            end
        end
    end
    
    def apply_badconsequence(m)
        
        badconsequence = m.bc
        
        n_levels = badconsequence.someLevels
        
        decrement_levels(n_levels)
       
        @pending_bc = badconsequence.adjust_to_fit_treasure_list(@visible_treasures, @hidden_treasures)
        
        set_pending_badconsequence(@pending_bc)
    end
    
    def can_make_treasure_visible(t)
        
        can = true;
        
        if (@visible_treasures.length < 5) 
        
            type = t.type;
            
            case (type) 
            
            when TreasureKind::ONEHAND 
            
                dos_manos = false
            
                @visible_teasures.each do |tr|
             
                    if (tr.type == TreasureKind::BOTHHANDS) 
                        can = false
                        dos_manos = true
                    end
                end
            
                if (dos_manos == false)            
             
                    i = 0
                    
                    @visible_treasures.each do |tr|
                
                        if (tr.type == TreasureKind::ONEHAND) 
                            i = i+1
                        end
                    end

                    if (i == 2) 
                        can = false 
                    end   
            
                else 
                    @visible_treasures.each do |tr|
                        
                        if (tr.type == type) 
                            can = false
                        end
                    end
                end    
            end
        end
          
        return can      
    end
   
    def how_many_visible_treasures(t_kind)
 
        i = 0
        @visible_treasures.each do |t|
 
            if t.type == t_kind then
                i = i + 1
            end
        end 
        
        return i       
    end
    
    def die_if_no_treasures
        
        if @visible_treasures.empty? && @hidden_treasures.empty?
            @dead = true
        end
    end
    
    def give_me_a_treasure
        @random = rand(CardDealer.unused_treasures.size)
        
        return CardDealer.unused_treasures.get(random)
    end
    
    def combat(m)
       
        my_level = self.get_combat_level
        monster_level = m.combatLevel

        if (my_level > monster_level)
        
            apply_prize(m)
       
            if (self.level >= @@MAXLEVEL) 
                combatResult = CombatResult::WINGAME
                
            else 
                combatResult = CombatResult::WIN 
            end
  
        else
            apply_badconsequence(m)
       
            if (should_convert())
                combatResult = CombatResult::LOSEANDCONVERT;
            else
                combatResult = CombatResult::LOSE;
            end
      
            combatResult = CombatResult::LOSE;
        end
    
        return combatResult
    end
    
    def make_treasure_visible(t)
        
        if(can_make_treasure_visible(t))
            @visible_treasures << t
            
            @hidden_treasures.delete(t)
        end
    end
    
    def discard_visible_treasure(t)
        @visible_treasures.delete(t)
        
        if(@pending_bc != nil && @pending_bc.is_empty)
            @pending_bc.substract_visible_treasure(t)
        end
        
        die_if_no_treasures
    end
    
    def discard_hidden_treasure(t)
        @hidden_treasures.delete(t)
        
        if(@pending_bc != nil && @pending_bc.is_empty)
            @pending_bc.substract_hidden_treasure(t)
        end
        
        die_if_no_treasures
    end
    
    def steal_treasure
        
        treasure = nil
        
        if(@canISteal)
            
            if(can_you_give_me_a_treasure)
                treasure = @enemy.give_me_a_treasure
                
                @hidden_treasures.add(treasure)
                
                self.have_stolen
            end
        end
        
        return treasure
    end
    
    def discard_all_treasures
        
        visible_copy = Array.new(@visible_treasures)
        hidden_copy = Array.new(@hidden_treasures)
        
        visible_copy.each do |t|
            discard_visible_treasure(t)
        end
        
        hidden_copy.each do |t|
            discard_hidden_treasure(t)
        end
    end
    
    def can_you_give_me_a_treasure
        return ((!@visible_treasures.empty?) && !(@hidden_treasures.empty?))
    end
    
    def have_stolen
        @canISteal = false
    end
    
    def init_treasures
        
        dealer = CardDealer.instance
        
        dice = Dice.instance
        
        bring_to_life
        
        treasure = dealer.next_treasure
        @hidden_treasures << treasure
        
        number = dice.next_number
        
        if (number > 1)
            treasure = dealer.next_treasure
            @hidden_treasures << (treasure)
        end
        
        if (number == 6)
            treasure = dealer.next_treasure
            @hidden_treasures << (treasure)
        end
    end
    
    def valid_state
        return (@pending_bc.is_empty && @hidden_treasures.length <= 4)
    end

    def should_convert
        should = false
        
        dice = Dice.instance
        
        if(dice.next_number == 1)
            should = true
        end
        
        return should
    end
    
end
end

File: P_5/Napakalaki_Java_P5/src/napakalaki/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;
import java.util.Random;
import GUI.Dice;

/**
 *
 * @author ixjosemi
 */

public class Player {

    private final int MAXLEVEL = 10;

    private String name;
    private int level = 1;
    private boolean dead = true;
    public boolean canISteal = true;

    public BadConsequence pendingBc = new NumericBadConsequence("",0,0,0);
    private Player enemy;
    
    protected ArrayList <Treasure> visibleTreasures = new ArrayList();
    protected ArrayList <Treasure> hiddenTreasures = new ArrayList();

    // Constructor
    public Player(String name){
        this.name = name;
    }

    // Constructor de copia
    public Player (Player p){

        this.name = p.name;
        this.level = p.level;
        this.dead = p.dead;
        this.canISteal = p.canISteal;
        this.enemy = p.enemy;
        this.pendingBc = p.pendingBc;
        
        this.visibleTreasures = p.visibleTreasures;
        this.hiddenTreasures = p.hiddenTreasures;
    }

    // Devuelve el nombre del jugador
    public String getName(){
        return this.name;
    }

    // Devuelve el enemigo del jugador
    public Player getEnemy(){
        return enemy;
    }

    // Devuelve a la vida
    private void bringToLife(){
        dead = false;
    }

    // Devuelve el nivel de combate del jugador, añadiendole el bonus de los 
    // tesoros visibles
    public int getCombatLevel(){

        int combatLevel = level;

        for(Treasure t: visibleTreasures)
            combatLevel = combatLevel + t.getBonus();

        return combatLevel;
    }

    // Devuelve el nivel del jugador
    public int getLevels(){
        return level;
    }

    // Devuelve el mal rollo pendiente del jugador
    public BadConsequence getPendingBadConsequence(){
        return pendingBc;
    }
    
    // Devuelve los tesoros ocultos del jugador
    public ArrayList <Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }

    // Devuelve los tesoros visible del jugador
    public ArrayList <Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }

    // Devuelve si el jugador actual esta muerto o no
    public boolean isDead(){
        return this.dead;
    }
    
    // Devuelve si el jugador actual es sectario o no
    public String IsCultist(){
        return "No";
    }

    // Aumenta los niveles del jugador
    private void incrementLevels(int levels){
        this.level += levels;

    }

    // Reduce los niveles del jugador
    private void decrementeLevels(int levels){

        this.level -= levels;

        if (this.level < 1)
            this.level = 1;
    }
    
    // Asigna el mal rollo al jugador
    private void setPendingBadConsequence(BadConsequence bc){
        this.pendingBc = bc;
    }
    
    // Aplica el buen rollo del monstruo al que el jugador ha vencido, sumandole
    // los niveles que le corresponden y pidiendo los tesoros que el buen rollo 
    // indique, los cuales se añaden a los tesoros ocultos.
    private void applyPrize(Monster m){
        
        int nLevels = m.getLevelsGained();
        
        incrementLevels(nLevels);
        
        int nTreasures = m.getTreasuresGained();
        
        if(nTreasures > 0){
            
            CardDealer dealer = CardDealer.getInstance();    
            Treasure treasure;
    
            for(int i = 0; i < nTreasures; i++){
                treasure = dealer.nextTreasure();
                hiddenTreasures.add(treasure);
            }       
        }
    }
    
    // Aplica el mal rollo al jugador al perder contra un monstruo, restandole
    // los niveles que correspondan y guardando una copia de un objeto
    // badconsequence ajustada a los tesoros que se pueden perder
    private void applyBadConsequence(Monster m){
        
       BadConsequence badconsequence = m.getBadConsequence();
       
       int nLevels = badconsequence.getLevels();
       
       decrementeLevels(nLevels);
       
       BadConsequence pendingBad =
               badconsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
       
       setPendingBadConsequence(pendingBc);
       
    }
          
    // Comprueba si el tesoro que le pasemos se puede pasar a visible segun
    // las reglas del juego
    private boolean canMakeTreasureVisible(Treasure t){
        
        boolean result = false;
        int how_many = howManyVisibleTreasures(t.getType());
        
        if(t.getType() == TreasureKind.ONEHAND){
            if(howManyVisibleTreasures(TreasureKind.BOTHHANDS) == 0){ // Si no tiene equipado una dos manos
                if(how_many == 2)
                        result = false;
            }
            else{
                result = false;
            }
        }
        
        else if(t.getType() == TreasureKind.BOTHHANDS){
            if(how_many == 1 || (howManyVisibleTreasures(TreasureKind.ONEHAND) > 0)) // Si tiene ya equipada una o tiene ya una "una mano"
                result = false;
        }
        
        else{
            if(how_many > 0)
                result = false;
        }
                
        return result;
    }
    
    // Devuelve el numero de tesoros visibles que tiene el jugador
    private int howManyVisibleTreasures(TreasureKind tKind){
        
        int i = 0;
        
        for(Treasure t  : this.visibleTreasures){
            
            if (t.getType() == tKind)
                i++;
        }
        return i;
    }
    
    // Cambia el estado del jugador a "muerto" si este ha perdido todos sus 
    // tesoros
    private void dieIfNoTreasure(){
        if(visibleTreasures.isEmpty() && hiddenTreasures.isEmpty())
            dead = true;
    }
     
    // Combate contra el monstruo actual, si el nivel del jugador es mayor que
    // el nivel del monstruo, se aplica el buen rollo y se puede ganar el combate
    // o el juego, en caso contrario, el jugador pierde el combate y se le aplica
    // el mal rollo
    public CombatResult combat(Monster m){
        
        CombatResult combatResult;
        
        int myLevel = getCombatLevel();
        int monsterLevel = this.getOponentLevel(m);
        
        if(!canISteal){
            Dice dice = Dice.getInstance();
            
            int number = dice.nextNumber();
            
            if(number < 3){
                int enemyLevel = enemy.getCombatLevel();
                
                monsterLevel += enemyLevel;
            }
        }
        
        if(myLevel > monsterLevel){
            applyPrize(m);
            
            if(level >= MAXLEVEL)
                combatResult = CombatResult.WINGAME;
           
            else
                combatResult = CombatResult.WIN;
        }
        else{
            applyBadConsequence(m);

            if (this.ShouldConvert())
                combatResult = CombatResult.LOSEANDCONVERT;
            else
                combatResult = CombatResult.LOSE;
        }
        return combatResult;
    }
    
    // Pasa un tesoro de oculto a visible
    public void makeTreasureVisible(Treasure t){
        
        if(canMakeTreasureVisible(t)){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    // Descarta un tesoro t que sea visible
    public void discardVisibleTreasure(Treasure t){
        
        visibleTreasures.remove(t);
        
        if(pendingBc != null && !(pendingBc.isEmpty()))
            pendingBc.substractVisibleTreasures(t);
        
        CardDealer dealer = CardDealer.getInstance();
        
        dealer.giveTreasureBack(t);
        
        dieIfNoTreasure();
    }
    
    // Descarta un tesoro t que sea oculto
    public void discardHiddenTreasure(Treasure t){
        
        hiddenTreasures.remove(t);
        
        if(pendingBc != null && !(pendingBc.isEmpty()))
            pendingBc.substractHiddenTreasures(t);
        
        CardDealer dealer = CardDealer.getInstance();
        
        dealer.giveTreasureBack(t);
        
        this.dieIfNoTreasure();
    }
    
    // Devuelve true en caso de que el jugador no tenga un mal rollo que cumplir
    // y no tiene mas de 4 tesoros ocultos
    public boolean validState(){
        return pendingBc.isEmpty() && hiddenTreasures.size() <= 4;
    }
    
    // Cuando un jugador se encuentra en su primer turno o se queda sin tesoros, 
    // hay que proporcionarle nuevos tesoros para que pueda continuar jugando
    // El numero de tesoros proporcionado viene dado por el valor que obtenga al
    // tirar el dado
    public void initTreasures(){
        
        CardDealer dealer = CardDealer.getInstance();
      
        Dice dice = Dice.getInstance();
        
        bringToLife();
        
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        
        int number = dice.nextNumber();
        
        if (number > 1){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        } 
        
        if (number == 6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
    }
    
    // Cuando el jugador quiere robar un tesoro, este metodo comprueba si es 
    // posible hacerlo y que su enemigo tenga tesoros ocultos que puedan
    // ser robados. El jugador no puede volver a robar tesoros durante el resto
    // de la partida. En el caso de no poder robar un tesoro se notifica.
    public Treasure stealTreasure(){
        
        Treasure treasure = null;
        
        if(canISteal){
            
            if(enemy.canYouGiveMeATreasure()){
                treasure = enemy.giveMeATreasure();
                
                hiddenTreasures.add(treasure);
                
                haveStolen();
            }
        }       
        return treasure;
    }
    
    // Le da valor al atributo "enemy"
    public void setEnemy(Player enemy){
        this.enemy = enemy;
    }
    
    // Devuelve un tesoro aleatorio de entre todos los tesoros ocultos del jugador
    private Treasure giveMeATreasure(){
       
        Random rnd = new Random();
        
        int i = rnd.nextInt(hiddenTreasures.size());
        
        return(hiddenTreasures.remove(i));
    }
    
    // Devuelve si el jugador no ha robado ningun tesoro y "false" en caso 
    // de que si lo haya hecho
    public boolean canISteal(){
        return this.canISteal;
    }
    
    // Devuelve trye si el jugador tiene tesoros que pueden ser robados por
    // otros y "false" en caso de que no
    private boolean canYouGiveMeATreasure(){
        return !hiddenTreasures.isEmpty();
    }
    
    // Cambia CanISteal a false cuando el jugador roba un tesoro
    private void haveStolen(){
        canISteal = false;
    }
     
    // El jugador descarta todos sus tesoros, tanto visibles como ocultos
    public void discardAllTreasures(){
        
        ArrayList <Treasure> visibleCopy = new ArrayList(visibleTreasures);
        ArrayList <Treasure> hiddenCopy = new ArrayList(hiddenTreasures);
        
        for(Treasure t: visibleCopy)
            discardVisibleTreasure(t);
        
        for(Treasure t: hiddenCopy)
            discardHiddenTreasure(t);
    }
    
    // Devuelve el nivel del monstruo del oponente
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    // Lanza el dado y se devolvera "true" si se obtiene un 1 y "false" en caso
    // de que no
    protected boolean ShouldConvert(){
        boolean should = false;
        
        Dice dice = Dice.getInstance();
  
        if (dice.nextNumber()==1)
            should = true;
        
        return (should);
    }
    
    @Override
    public String toString(){
        return  "" + name;
    }
}
